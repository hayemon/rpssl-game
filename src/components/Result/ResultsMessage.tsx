import React from 'react'
import { Results } from 'types/PlayResult'
import styles from './index.module.scss'

type Props = {
  results: Results | undefined
}

const mapper: Record<Results, { text: string; className: string }> = {
  win: {
    text: 'You win!',
    className: styles.win,
  },
  lose: {
    text: 'You lose!',
    className: styles.lose,
  },
  tie: {
    text: 'Tie...',
    className: styles.tie,
  },
}

const ResultsMessage: React.FC<Props> = ({ results }) => {
  if (results === undefined) {
    return null
  }

  const mappedValue = mapper[results]

  return (
    <div className={`glass ${styles.box}`}>
      <h1 className={mappedValue.className}>{mappedValue.text}</h1>
    </div>
  )
}

export default ResultsMessage
