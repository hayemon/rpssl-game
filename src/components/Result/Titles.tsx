import React from 'react'
import styles from './index.module.scss'

const Titles: React.FC = () => {
  return (
    <div className={styles.row}>
      <h3 className={`${styles.title} ${styles.playerTitle}`}>Player</h3>
      <h3 className={`${styles.title} ${styles.computerTitle}`}>Computer</h3>
    </div>
  )
}

export default Titles
