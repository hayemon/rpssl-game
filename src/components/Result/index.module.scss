@import '../../mixins';
@include rotate-animation;

.container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 75vh;
  margin: 0 20vw 0 20vw;
}

.container h1 {
  position: absolute;
  font-size: 25rem;
  color: rgba(255, 255, 255, 0.1);
}

.row {
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  align-items: center;
  width: 100%;
}

@mixin box {
  width: 500px;
  height: 200px;
  border-radius: $border-radius;
  display: flex;
  transform-origin: center;
  perspective-origin: center;
  overflow: hidden;
}

.box {
  @include box;
  animation: rotate 5s infinite linear, fade-in 1s;
}

@keyframes fade-in {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

@keyframes wobble {
  0% {
    transform: perspective(600px) rotateX(14deg);
  }
  25% {
    transform: perspective(600px) rotateY(14deg);
  }
  50% {
    transform: perspective(600px) rotateX(-14deg);
  }
  75% {
    transform: perspective(600px) rotateY(-14deg);
  }
  100% {
    transform: perspective(600px) rotateX(14deg);
  }
}

.spinner {
  @include box;
  width: 150px;
  height: 150px;
  animation: wobble 5s infinite linear;
}

@mixin burst-in($animation-name, $initial-position) {
  @keyframes #{$animation-name} {
    0% {
      transform: translateY($initial-position);
      opacity: 1;
    }
    20% {
      transform: translateY(0);
      opacity: 0;
    }
    25% {
      opacity: 0.5;
    }
    30% {
      opacity: 0;
    }
    35% {
      opacity: 0.8;
    }
    40% {
      opacity: 0;
    }
    100% {
      opacity: 1;
    }
  }
}

@include burst-in(player-animation, 200px);
@include burst-in(computer-animation, -200px);

.player {
  // animation: player-animation 1s;
}

.computer {
  // animation: computer-animation 1s;
}

.title {
  color: rgba(255, 255, 255, 0.3);
  letter-spacing: 5px;
  display: flex;
  width: 100%;
}

.playerTitle {
  justify-content: end;
  padding-right: 1rem;
}

.playerTitle::after {
  content: '';
  border-right: 1px solid rgba(255, 255, 255, 0.2);
  position: absolute;
  left: 0;
  height: 230px;
  width: 50vw;
}

.computerTitle {
  justify-content: start;
  padding-left: 1rem;
}

.spinner::before {
  @include spinner-before;
}

.spinner::after {
  @include spinner-after;
}

.spinner img {
  @include img-in-box;
}

.box h1 {
  width: 100%;
  height: 100%;
  z-index: 10;
  text-align: center;
  font-size: 6em;
  text-transform: uppercase;
  outline: none;
  color: white;
  line-height: 0.75;
}

.win {
  @include text-glow($win-color);
}

.lose {
  @include text-glow($lose-color);
}
