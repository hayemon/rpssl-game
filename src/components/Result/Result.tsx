import useCountdown from 'hooks/useCountdown'
import React from 'react'
import { selectPlayerChoice, selectResults } from 'store/selectors'
import { useGetChoiceQuery } from 'store/slices/apiSlice'
import urlMapper from 'utils/urlMapper'
import styles from './index.module.scss'
import ResultsMessage from './ResultsMessage'
import Titles from './Titles'

const Result: React.FC = () => {
  const playerChoice = selectPlayerChoice()
  const results = selectResults()

  const { data } = useGetChoiceQuery(playerChoice, {
    skip: playerChoice === undefined,
  })

  const count = useCountdown(playerChoice)

  return (
    <div className={styles.container}>
      {playerChoice !== undefined && !results && <Titles />}
      {playerChoice !== undefined && !results && (
        <div className={styles.row}>
          <div className={`glass ${styles.spinner} ${styles.player}`}>
            <img src={urlMapper[playerChoice]} />
          </div>
          {data && (
            <div className={`glass ${styles.spinner} ${styles.computer}`}>
              <img src={urlMapper[data.name]} />
            </div>
          )}
        </div>
      )}
      {count > 0 && <h1>{count}</h1>}
      <ResultsMessage results={results} />
    </div>
  )
}

export default Result
