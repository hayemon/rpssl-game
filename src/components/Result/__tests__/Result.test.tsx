import '@testing-library/jest-dom/extend-expect'
import { renderWithProviders } from 'store/test/testUtils'
import Result from '../Result'

describe('Result', () => {
  it('renders', () => {
    const { asFragment } = renderWithProviders(<Result />)
    expect(asFragment()).toMatchSnapshot()
  })
})
