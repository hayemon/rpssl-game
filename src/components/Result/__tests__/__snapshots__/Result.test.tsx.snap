// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result renders 1`] = `
<DocumentFragment>
  <div>
    <div>
      <h3
        class="undefined undefined"
      >
        Player
      </h3>
      <h3
        class="undefined undefined"
      >
        Computer
      </h3>
    </div>
    <div>
      <div
        class="glass undefined undefined"
      >
        <img
          src="test-file-stub"
        />
      </div>
    </div>
    <h1>
      3
    </h1>
  </div>
</DocumentFragment>
`;
